nfs: implement ControllerModifyVolume to change NFS-server hostname

nixpanic opened the pull request:
The new hostname of the NFS-server is stored in a journal entry for the
volume. The new NFS-server will only be used the next time the NFS
Volume is mounted on a worker node.

It is not possible to update the VolumeContext of a Volume, so the old
hostname of the NFS-server will be stored there forever.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic committed:
nfs: create helpers for storing volume attributes in the journal

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
nfs: fail early in NodeGetVolumeStats

All functions use a fail-early-continue-on-success flow,
NodeGetVolumeStats was not following this.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
nfs: move NFSVolume to shared types package

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
nfs: make `paramServer` shared option public

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
nfs: implement ControllerModifyVolume to change NFS-server hostname

The new hostname of the NFS-server is stored in a journal entry for the
volume. The new NFS-server will only be used the next time the NFS
Volume is mounted on a worker node.

It is not possible to update the VolumeContext of a Volume, so the old
hostname of the NFS=server will be stored there forever.

Closes: #5420
Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 reviewed (commented):
@nixpanic is this tested code? i dont see ControllerServiceCapability_RPC_MODIFY_VOLUME capability is exposed for nFS

nixpanic commented:
> @nixpanic is this tested code? i dont see ControllerServiceCapability_RPC_MODIFY_VOLUME capability is exposed for nFS

Still trying to add some e2e tests, the csi-provisioner and csi-resizer sidecars have just gained the ability to pass credentials to the `ControllerModifyVolume` request.

nixpanic committed:
e2e: cleanup k8s.io module dependencies

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
e2e: add VolumeAttributesClass test for NFS

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.34/nfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/nfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
> @nixpanic is this tested code? i dont see ControllerServiceCapability_RPC_MODIFY_VOLUME capability is exposed for nFS

Verified :stuck_out_tongue_winking_eye:  that it fails with this:
```
  I1212 12:26:38.595866 75590 pvc.go:85] PVC cephcsi-nfs-pvc Event: ProvisioningFailed - CSI driver does not support VolumeAttributesClass: controller MODIFY_VOLUME capability is not reported
```

nixpanic committed:
doc: include reference to NFS/ControllerModifyVolume in release-notes

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic committed:
testing: use canary images for csi-resizer/provisioner

Only the canary version suppors the VolumeAttributesClass with secrets
in the ControllerModifyVolume gRPC.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic committed:
deploy: allow NFS-provisioner to access volumeattributesclasses

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
Still trying to use the old nfs-server:

```
  Warning  FailedMount  8s (x13 over 10m)  kubelet            MountVolume.SetUp failed for volume "pvc-c64d91d7-9097-46f3-9594-eb993e5851b7" : rpc error: code = Internal desc = nfs: failed to mount "relocated.example.net:/0001-0024-e044d94f-5dfb-4f7b-a7e6-44895f196074-0000000000000001-38cbb13b-1f15-4a9b-80ff-3017574691a1" to "/var/lib/kubelet/pods/4ebe7989-0889-4af1-b138-48c67a90debd/volumes/kubernetes.io~csi/pvc-c64d91d7-9097-46f3-9594-eb993e5851b7/mount" : mount failed: exit status 32
Mounting command: mount
Mounting arguments: -t nfs relocated.example.net:/0001-0024-e044d94f-5dfb-4f7b-a7e6-44895f196074-0000000000000001-38cbb13b-1f15-4a9b-80ff-3017574691a1 /var/lib/kubelet/pods/4ebe7989-0889-4af1-b138-48c67a90debd/volumes/kubernetes.io~csi/pvc-c64d91d7-9097-46f3-9594-eb993e5851b7/mount
Output: mount.nfs: Failed to resolve server relocated.example.net: Name or service not known
 stderr: ""
```

nixpanic commented:
It seems I missed handling of `mutable_parameters` in `CreateVolume` CSI procedure.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic committed:
ci: include publish-secret in StorageClass for NFS

The NFS node-plugin does not support staging, it only uses publishing.
For VolumeAttributeClass / ControllerModifyVolume the mutable parameters
are stored in the Ceph backend. This means that during volume publishing
the node-plugin needs to get the updated parameters from the Ceph
cluster (hence the publish secret requirement).

/tmp/csi has been added to the node-plugin Pod so that the temporary
credentials file can be written.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32/cephfs

Rakshith-R commented on examples/nfs/storageclass.yaml:
```diff
@@ -38,8 +38,9 @@ parameters:
   csi.storage.k8s.io/provisioner-secret-namespace: default
   csi.storage.k8s.io/controller-expand-secret-name: csi-cephfs-secret
   csi.storage.k8s.io/controller-expand-secret-namespace: default
-  csi.storage.k8s.io/node-stage-secret-name: csi-cephfs-secret
-  csi.storage.k8s.io/node-stage-secret-namespace: default
```
Let's preserve node-stage-secret for future use ?

nixpanic replied:
There is no `NodeStageVolume` call for NFS. I think there no need to refer to a secret either?

Rakshith-R commented on internal/nfs/controller/controllerserver.go:
```diff
@@ -119,6 +120,21 @@ func (cs *Server) CreateVolume(
+	if req.GetMutableParameters() != nil {
+		log.DebugLog(ctx, "modifying parameters: %v", req.GetMutableParameters())
+
+		_, err = cs.ControllerModifyVolume(ctx,
```
This currently works since 
- we don't have Locks in NFS controllerserver.
- currently controllerModify for nfs does not need to modify underlying cephfs attributes.

We may face issues in the future if we decide to change either of the behaviors mentioned above ?

nixpanic replied:
Yes, that could possibly be an issue. But in case the points you mention get implemented, this can be refactored as well (the code is quite small and not complex. I think).

nixpanic commented:
@Madhu-1 , @Rakshith-R care to have an other look? There is a e2e test included now, and it (finally) works.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34/cephfs

nixpanic committed:
doc: mention support for NFS-server hostname changing

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Rakshith-R commented on build.env:
```diff
@@ -61,10 +61,11 @@ ROOK_VERSION=v1.16.4
 CSI_SNAPSHOTTER_VERSION=v8.4.0
-CSI_RESIZER_VERSION=v2.0.0
-CSI_PROVISIONER_VERSION=v6.0.0
+CSI_RESIZER_VERSION=canary
+CSI_PROVISIONER_VERSION=canary
```
Can we have a TODO and tracker to revert this to released version ? 
(:sweat: we made a similar change and forgot to revert before too)

nixpanic replied:
sure, created #5955 for that.

Rakshith-R reviewed (dismissed):
Just a small nit.

Madhu-1 commented on README.md:
```diff
@@ -114,6 +114,7 @@ for its support details.
 |        | Creating and deleting snapshot                            | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=v16.2.0)  | >= v1.17.0         |
 |        | Provision volume from snapshot                            | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=v16.2.0)  | >= v1.17.0         |
 |        | Provision volume from another volume                      | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=v16.2.0)  | >= v1.16.0         |
+|        | Modify the NFS-server hostname with VolumeAttributeClass  | Alpha          | >= v3.17.0         | >= v1.12.0       | Pacific (>=v16.2.0)  | >= v1.34.0         |
```
Can we mark this as Modify volume instead of adding details because this will keep of exapanding if we add details what we do in modify volume

nixpanic replied:
Sure, done.

Madhu-1 commented on e2e/nfs.go:
```diff
@@ -178,8 +178,8 @@ func createNFSStorageClass(
 
-	sc.Parameters["csi.storage.k8s.io/node-stage-secret-namespace"] = cephCSINamespace
-	sc.Parameters["csi.storage.k8s.io/node-stage-secret-name"] = cephFSNodePluginSecretName
+	sc.Parameters["csi.storage.k8s.io/node-publish-secret-namespace"] = cephCSINamespace
```
@nixpanic what about the upgraded cluster? Can we document the steps for the upgraded cluster as well, where they already have nfs pvc?

nixpanic replied:
Yeah, I'll put something together based on https://github.com/kubernetes-csi/docs/pull/631

Madhu-1 commented on e2e/nfs.go:
```diff
@@ -227,6 +227,79 @@ func createNFSStorageClass(
+	f *framework.Framework,
+	params map[string]string,
+) error {
+	vacPath := fmt.Sprintf("%s/%s", nfsExamplePath, "vac-relocated.yaml")
```
can we keep this file name as volume.yaml instead of relocated as it will be generic and can be expanded later for new fields

nixpanic replied:
Sure, will do.

Madhu-1 commented on internal/nfs/controller/controllerserver.go:
```diff
@@ -185,3 +201,37 @@ func (cs *Server) DeleteSnapshot(
 	return cs.backendServer.DeleteSnapshot(ctx, req)
 }
+
+func (cs *Server) ControllerModifyVolume(
```
Add a comment to this new function?

Madhu-1 commented on internal/nfs/controller/controllerserver.go:
```diff
@@ -185,3 +201,37 @@ func (cs *Server) DeleteSnapshot(
+	req *csi.ControllerModifyVolumeRequest,
+) (*csi.ControllerModifyVolumeResponse, error) {
+	secret := req.GetSecrets()
+	cr, err := util.NewAdminCredentials(secret)
```
lets use `NewUserCredentials` because adminCreds are deprecated

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -282,3 +302,37 @@ func getSource(volContext map[string]string) (string, error) {
+		// no secrets, continue as if not set in metadata
+		return "", nil
+	}
+	cr, err := util.NewAdminCredentials(secrets)
```
use `NewUserCredentials` credentials

Madhu-1 reviewed (commented):
small nits, LGTM

nixpanic committed:
doc: add details about VolumeAttributeClass secrets

Only Kubernetes 1.34 and newer support VolumeAttributeClass
functionality (GA). Ceph-CSI also depends on certain versions of the
Kubernetes CSI external-provisioner and external-resizer.

See-also: kubernetes-csi/docs#631
Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
@Madhu-1 , please have a look again, all comments should have been addressed.

Madhu-1 commented on docs/volumeattributeclass.md:
```diff
@@ -0,0 +1,66 @@
+## Adding support to existing Volumes
+
+PersistentVolumeClaims that have not been created with the right secret
+references in the StorageClass will not be modifyable with a
```
modifyable to modifiable

Madhu-1 commented on docs/volumeattributeclass.md:
```diff
@@ -0,0 +1,66 @@
+# VolumeAttributeClass for Volume Modification
```
VolumeAttributeClass to VolumeAttributesClass in all the places and also update the file name here and in example

nixpanic replied:
argh, yes, a common typo :disappointed:

Madhu-1 reviewed (commented):
small doc nits

Madhu-1 commented on docs/volumeettributesclass.md:
```diff
@@ -0,0 +1,66 @@
+# VolumeAttributesClass for Volume Modification
+
+Kubernetes offers a method for modifying Volume parameters after they have been
+created. This is done through VolumeAttributeClasses and is described in a
```
`VolumeAttributeClasses` to `VolumeAttributesClasses`

Madhu-1 commented on docs/volumeettributesclass.md:
```diff
@@ -0,0 +1,66 @@
+
+## Prerequisites
+
+- Kubernetes 1.34 is the first release where support for VolumeAttributeClasses
```
VolumeAttributesClasses

Madhu-1 commented on examples/nfs/volumeattributesclass.yaml:
```diff
@@ -0,0 +1,11 @@
+---
```
The file name needs to be updated as well

Madhu-1 reviewed (commented):
almost there

iPraveenParihar commented on internal/nfs/controller/volume.go:
```diff
@@ -296,18 +296,18 @@ func (nv *NFSVolume) getNFSCluster() (string, error) {
-	clusterName, err := j.FetchAttribute(nv.ctx, mdPool, nv.objectUUID, clusterNameKey)
+	value, err := j.FetchAttribute(nv.ctx, mdPool, nv.objectUUID, key)
 	if err != nil && errors.Is(err, util.ErrPoolNotFound) || errors.Is(err, util.ErrKeyNotFound) {
 		return "", fmt.Errorf("cluster name for %q %w: %w", nv.objectUUID, ErrNotFound, err)
```
error should be generalised here too
```
failed to get attribute %s for %q %w: %w", key, nv.objectUUID, ErrNotFound, err)
```

iPraveenParihar commented on internal/nfs/controller/volume.go:
```diff
@@ -334,10 +334,20 @@ func (nv *NFSVolume) setNFSCluster(clusterName string) error {
-	err = j.StoreAttribute(nv.ctx, mdPool, nv.objectUUID, clusterNameKey, clusterName)
+	err = j.StoreAttribute(nv.ctx, mdPool, nv.objectUUID, key, value)
 	if err != nil {
 		return fmt.Errorf("failed to store cluster name: %w", err)
```
here too, `failed to store attribute %s: %w", err`

iPraveenParihar commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -283,3 +302,37 @@ func getSource(volContext map[string]string) (string, error) {
+	}
+	cr, err := util.NewUserCredentials(secrets)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to retrieve admin credentials: %v", err)
```
```suggestion
		log.ErrorLog(ctx, "failed to retrieve user credentials: %v", err)
```

iPraveenParihar reviewed (commented):
nits

nixpanic commented:
Thanks again for the reviews! Hopefully I didn't miss anything this time :crossed_fingers:

nixpanic commented:
@mergifyio queue

nixpanic commented:
@mergifyio requeue

nixpanic commented:
@mergifyio requeue

nixpanic commented:
@mergifyio requeue

nixpanic commented:
@mergifyio requeue

iPraveenParihar commented:
@Mergifyio requeue

nixpanic commented:
@mergifyio requeue

nixpanic commented:
@mergifyio requeue

nixpanic commented:
It seems there has been an update to the `csi-resizer:canary` image and `watch` is required now too:

```
  E0201 19:25:29.550775       1 reflector.go:205] "Failed to watch" err="volumeattributesclasses.storage.k8s.io is forbidden: User \"system:serviceaccount:cephcsi-e2e-ae4861ac:nfs-csi-provisioner\" cannot watch resource \"volumeattributesclasses\" in API group \"storage.k8s.io\" at the cluster scope" logger="UnhandledError" reflector="k8s.io/client-go/informers/factory.go:160" type="*v1.VolumeAttributesClass"
```

Updated the RBAC for the provisioner, please have a look again.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
strange, it looks like these StorageClass keys are not accepted, but those were added to the external-provisioner :thinking: 

- csi.storage.k8s.io/controller-modify-secret-name
- csi.storage.k8s.io/controller-modify-secret-namespace

The logs show
```
  I0203 13:29:35.730236       1 csi-provisioner.go:170] "Version" version="v6.0.0"
```

Not sure that it is the canary image mentioned in `build.env`. Even stranger is that this has passed CI before, and inspection of the logs showed correct functioning.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
> Not sure that it is the canary image mentioned in `build.env`. Even stranger is that this has passed CI before, and inspection of the logs showed correct functioning.



> strange, it looks like these StorageClass keys are not accepted, but those were added to the external-provisioner 🤔
> 
>     * csi.storage.k8s.io/controller-modify-secret-name
> 
>     * csi.storage.k8s.io/controller-modify-secret-namespace
> 
> 
> The logs show
> 
> ```
>   I0203 13:29:35.730236       1 csi-provisioner.go:170] "Version" version="v6.0.0"
> ```
> 
> Not sure that it is the canary image mentioned in `build.env`. Even stranger is that this has passed CI before, and inspection of the logs showed correct functioning.



> strange, it looks like these StorageClass keys are not accepted, but those were added to the external-provisioner 🤔
> 
>     * csi.storage.k8s.io/controller-modify-secret-name
> 
>     * csi.storage.k8s.io/controller-modify-secret-namespace
> 
> 
> The logs show
> 
> ```
>   I0203 13:29:35.730236       1 csi-provisioner.go:170] "Version" version="v6.0.0"
> ```
> 
> Not sure that it is the canary image mentioned in `build.env`. Even stranger is that this has passed CI before, and inspection of the logs showed correct functioning.

Added the canary image in commit "Testing: use canary images for csi-resizer/provisioner", not only in `build.env`, but also under `deploy/..../nfs`.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
> /test ci/centos/mini-e2e/k8s-1.34

```
  I0204 12:08:04.408826 74903 pvc.go:85] PVC csi-cephfs-pvc Event: ProvisioningFailed - Volume provisioning failed with infeasible error. Retries will be delayed. rpc error: code = InvalidArgument desc = failed to fetch monitor list using clusterID (clusterID-1): missing configuration for cluster ID: "clusterID-1"
```

https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e_k8s-1.34/460/display/redirect

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.35

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.35

nixpanic commented:
@Madhu-1 , NFS (and CephFS) uses admin credentials everywhere. Just using user credentials for ControllerModifyVolume and NodePublishVolume isn't straight forward. Reverted to admin credentials again, so that things at least work again.

Madhu-1 commented:
@nixpanic lets address the merge conflict and have a tracker to address the admin creds

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.35

nixpanic commented:
aaah! after renaming the VAC example, there was still a typo somewhere :cry: 

```
failed to create NFS voluemattributesclass: open ../examples/nfs//volumeattributeclass.yaml: no such file or directory
```

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.35

nixpanic commented:
@Madhu-1 and @Rakshith-R , this is now really ready for approval. The CI job is still running, but the NFS tests have passed.

nixpanic commented:
> @nixpanic lets address the merge conflict and have a tracker to address the admin creds

Tracked by #6026.

Rakshith-R commented:
@mergifyio requeue
